src/Storages/ExternalDataSourceConfiguration.h

```diff
@@ -0,0 +1,105 @@
+#pragma once
+
+#include <Parsers/ASTIdentifier.h>
+#include <Interpreters/Context.h>
+#include <Poco/Util/AbstractConfiguration.h>
+
+
+namespace DB
+{
+
+struct ExternalDataSourceConfiguration
+{
+    String host;
+    UInt16 port = 0;
+    String username;
+    String password;
+    String database;
+    String table;
+    String schema;
+
+    std::vector<std::pair<String, UInt16>> addresses; /// Failover replicas.
+
+    String toString() const;
+
+    void set(const ExternalDataSourceConfiguration & conf);
+};
+
+using ExternalDataSourceConfigurationPtr = std::shared_ptr<ExternalDataSourceConfiguration>;
+
+
+struct StoragePostgreSQLConfiguration : ExternalDataSourceConfiguration
+{
+    String on_conflict;
+};
+
+
+struct StorageMySQLConfiguration : ExternalDataSourceConfiguration
+{
+    bool replace_query = false;
+    String on_duplicate_clause;
+};
+
+struct StorageMongoDBConfiguration : ExternalDataSourceConfiguration
+{
+    String collection;
+    String options;
+};
+
+
+using EngineArgs = std::vector<std::pair<String, DB::Field>>;
+
+/* If storage engine's configuration was define via named_collections,
+ * return all options in ExternalDataSource::Configuration struct.
+ *
+ * Also check if engine arguments have key-value defined configuration options:
+ * ENGINE = PostgreSQL(postgresql_configuration, database = 'postgres_database');
+ * In this case they will override values defined in config.
+ *
+ * If there are key-value arguments apart from common: `host`, `port`, `username`, `password`, `database`,
+ * i.e. storage-specific arguments, then return them back in a set: ExternalDataSource::EngineArgs.
+ */
```

Comment:
```suggestion
/* If there is a storage engine's configuration specified in the named_collections,
 * this function returns valid for usage ExternalDataSourceConfiguration struct
 * otherwise std::nullopt is returned.
 *
 * If any configuration options are provided as key-value engine arguments, they will override
 * configuration values, i.e. ENGINE = PostgreSQL(postgresql_configuration, database = 'postgres_database');
 *
 * Any key-value engine argument except common (`host`, `port`, `username`, `password`, `database`)
 * is returned in EngineArgs struct.
 */
```